packages/transactions/src/blockhash.ts

```diff
@@ -34,3 +38,45 @@ export function assertIsBlockhash(putativeBlockhash: string): asserts putativeBl
         });
     }
 }
+
+export function setTransactionLifetimeUsingBlockhash<TTransaction extends BaseTransaction & IDurableNonceTransaction>(
+    blockhashLifetimeConstraint: BlockhashLifetimeConstraint,
+    transaction: TTransaction | (TTransaction & ITransactionWithSignatures)
+): Omit<TTransaction, keyof ITransactionWithSignatures | 'lifetimeConstraint'> & ITransactionWithBlockhashLifetime;
+export function setTransactionLifetimeUsingBlockhash<
+    TTransaction extends BaseTransaction | (BaseTransaction & ITransactionWithBlockhashLifetime)
+>(
+    blockhashLifetimeConstraint: BlockhashLifetimeConstraint,
+    transaction: TTransaction | (TTransaction & ITransactionWithSignatures)
+): Omit<TTransaction, keyof ITransactionWithSignatures> & ITransactionWithBlockhashLifetime;
```

Comment:
Method overloads! Check this out ([link](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAWxAG1gBzQTwBQCGAXIgM5QBOMYA5gDSIBGJYIyjAphQJQnlW0A3AChQkWAhTosuQizacKDZolbsuvVQq4ix0eElQYY2fMS3qKiAD5lK1ek3mXuiAN7DEXxDGCI8UDiYHHB+BIgAvFGIAOT8DjGuHt4piBQcUCAUSAQAdOnBBFB4jNwiKQC+iBxopBzunqle6ZnZiOEAVEzl3hXCfcJGMvgxaHBoMQwAjAAMZcIA9AuIAHoA-MKD0iayAEzTcyJLq2tAA)).

Essentially you can _completely_ change the TypeScript return type of a function, depending on its inputs.

Here we're saying ‘if you call this sucker with a durable nonce transaction then strip the `lifetimeConstraint`; otherwise just strip the `signatures`.’

I don't doubt there's a more compact way to achieve this, but this is where I landed.

Reply:
> And what does calling this with a durable nonce transaction do?

The goal here is to:

1. Preserve the type of the transaction passed in
2. …except strip the `lifetimeConstraint` if it's a durable nonce transaction and replace it with a blockhash one
3. …strip the signatures, if and only if the `lifetimeConstraint` changes

The typing around durable nonce transactions is to make sure that the type of `lifetimeConstraint` ends up as:

```
{ blockhash: Blockhash, lastValidBlockHeight: Slot }
```

…and not

```
{ blockhash: Blockhash, lastValidBlockHeight: Slot } | { nonce: Nonce }
```

To achieve that I had to _forcibly_ omit the `lifetimeConstraint` before putting it back via `& ITransactionWithBlockhashLifetime`.